File: config/src/test/scala/net/kemitix/thorp/config/ParseConfigFileTest.scala
package net.kemitix.thorp.config

import java.io.File
import java.nio.file.Paths
import java.util

import net.kemitix.thorp.filesystem.TemporaryFolder
import org.scalatest.FunSpec
import zio.DefaultRuntime

import scala.jdk.CollectionConverters._

class ParseConfigFileTest extends FunSpec with TemporaryFolder {

  private val empty = Right(ConfigOptions.empty)

  describe("parse a missing file") {
    val file = new File("/path/to/missing/file")
    it("should return no options") {
      assertResult(empty)(invoke(file))
    }
  }
  describe("parse an empty file") {
    it("should return no options") {
      withDirectory(dir => {
        val file = createFile(dir, "empty-file", List.empty[String].asJava)
        assertResult(empty)(invoke(file))
      })
    }
  }
  describe("parse a file with no valid entries") {
    it("should return no options") {
      withDirectory(dir => {
        val file = createFile(dir,
                              "invalid-config",
                              util.Arrays.asList("no valid = config items"))
        assertResult(empty)(invoke(file))
      })
    }
  }
  describe("parse a file with properties") {
    it("should return some options") {
      val expected = Right(
        ConfigOptions(
          List[ConfigOption](ConfigOption.Source(Paths.get("/path/to/source")),
                             ConfigOption.Bucket("bucket-name")).asJava))
      withDirectory(dir => {
        val file = createFile(dir,
                              "simple-config",
                              util.Arrays.asList("source = /path/to/source",
                                                 "bucket = bucket-name"))
        assertResult(expected)(invoke(file))
      })
    }
  }

  private def invoke(file: File) = {
    new DefaultRuntime {}.unsafeRunSync {
      ParseConfigFile.parseFile(file)
    }.toEither
  }
}

File: config/src/test/java/net/kemitix/thorp/config/ParseConfigFileTest.java
package net.kemitix.thorp.config;

import net.kemitix.thorp.filesystem.TemporaryFolder;
import org.assertj.core.api.WithAssertions;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.io.File;
import java.io.IOException;
import java.nio.file.Paths;
import java.util.Arrays;
import java.util.Collections;

public class ParseConfigFileTest
        implements WithAssertions, TemporaryFolder {

    @Test
    @DisplayName("when file is missing then no options")
    public void whenFileMissing_thenNoOptions() throws IOException {
        assertThat(invoke(new File("/path/to/missing/file")))
                .isEqualTo(ConfigOptions.empty());
    }
    @Test
    @DisplayName("when file is empty then no options")
    public void whenEmptyFile_thenNoOptions() {
        withDirectory(dir -> {
            File file = createFile(dir, "empty-file", Collections.emptyList());
            assertThat(invoke(file)).isEqualTo(ConfigOptions.empty());
        });
    }
    @Test
    @DisplayName("when no valid entried then no options")
    public void whenNoValidEntries_thenNoOptions() {
        withDirectory(dir -> {
            File file = createFile(dir, "invalid-config",
                    Arrays.asList("no valid = config items", "invalid line"));
            assertThat(invoke(file)).isEqualTo(ConfigOptions.empty());
        });
    }

    @Test
    @DisplayName("when file is valid then parse options")
    public void whenValidFile_thenOptions() {
        withDirectory(dir -> {
            File file = createFile(dir, "simple-config", Arrays.asList(
                    "source = /path/to/source",
                    "bucket = bucket-name"));
            assertThat(invoke(file)).isEqualTo(
                    ConfigOptions.create(
                            Arrays.asList(
                                    ConfigOption.source(Paths.get("/path/to/source")),
                                    ConfigOption.bucket("bucket-name"))));
        });
    }

    ConfigOptions invoke(File file) {
        try {
            return ParseConfigFile.parseFile(file);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }
}
